### PR TITLE
maximum memory stats

### PR DESCRIPTION
**Pull request by btiplitz**
this keeps track of the maximum memory used from the cache

### CODE REVIEW

**Comment by btiplitz** [comment]
> don't you mean current_memory_ ? Did you compile this?

It seems I did not have cuda on the build, so no. Working on adding cuda to my build now

File: src/cudamatrix/cu-allocator.cc [hunk 1]
```diff
@@ -223,6 +223,8 @@ void* CuMemoryAllocator::MallocFromSubregion(SubRegion *subregion,
   block->allocated = true;
   block->t = t_;
   allocated_block_map_[block->begin] = block;
+  allocated_memory_ += (block->end - block->begin);
+  if (allocated_memory_ > max_allocated_memory_) max_allocated_memory_ = allocated_memory_;
```
Comment (danpovey) [inline src/cudamatrix/cu-allocator.cc]: Sorry, a couple small comments.  Would you mind splitting this line?  It looks over the 80-char limit.

File: src/cudamatrix/cu-allocator.cc [hunk 2]
```diff
@@ -359,7 +361,9 @@ void CuMemoryAllocator::PrintMemoryUsage() const {
             << tot_time_taken_ << "/" << malloc_time_taken_
             << ", synchronized the GPU " << num_synchronizations_
             << " times out of " << (t_/2) << " frees; "
-            << "device memory info: " << GetFreeGpuMemory(NULL, NULL);
+            << "device memory info: " << GetFreeGpuMemory(NULL, NULL)
+            << "maximum requested: " << max_allocated_memory_
```
Comment (danpovey) [inline src/cudamatrix/cu-allocator.cc]: Here, please change 'requested' to 'allocated'.

File: src/cudamatrix/cu-allocator.cc [hunk 3]
```diff
@@ -413,6 +419,7 @@ void CuMemoryAllocator::Free(void *ptr) {
               << ptr;
   }
   MemoryBlock *block = iter->second;
+  allocated_memory_ -= (block->end-block->begin);
```
Comment (danpovey) [inline src/cudamatrix/cu-allocator.cc]: space around operator please (the '-')

File: src/cudamatrix/cu-allocator.h [hunk 4]
```diff
@@ -335,6 +335,11 @@ class CuMemoryAllocator {
   // this is only locked by the '*Locking' versions of the functions (necessary only
   // in multi-threaded applications).
   std::mutex mutex_;
+
+  // Keep track of the memory usage from the cache to track the maximum memory used by
+  //   the application
+  int max_allocated_memory_;
```
Comment (danpovey) [inline src/cudamatrix/cu-allocator.h]: please make this size_t, instead of int.  If int was 32-bit, this might not always be enough.

**Pull request merged**